openfisca_core/data_storage.py

```diff
@@ -1,136 +1,191 @@
 
-import numpy as np
+import numpy
 
+from openfisca_core import indexed_enums
 from openfisca_core import periods
-from openfisca_core.periods import ETERNITY
-from openfisca_core.indexed_enums import EnumArray
 
+from typing_extensions import Protocol
+
+
+class StorageLike(Protocol):
+
+    @abc.abstractmethod
+    def get(self, period: periods.Period) -> numpy.ndarray:
+        ...
+
+    @abc.abstractmethod
+    def put(self, value: numpy.ndarray, period: periods.Period) -> None:
+        ...
 
-class InMemoryStorage(object):
+    @abc.abstractmethod
+    def delete(self, period: Optional[periods.Period] = None) -> None:
+        ...
+
+    @abc.abstractmethod
+    def get_known_periods(self) -> List[periods.Period]:
+        ...
+
+    @abc.abstractmethod
+    def get_memory_usage(self) -> Dict[str, int]:
+        ...
+
+    def _pop(self, period: periods.Period, items: list) -> dict:
+        return {item: value for item, value in items if not period.contains(item)}
+
+
+class InMemoryStorage(StorageLike):
     """
-    Low-level class responsible for storing and retrieving calculated vectors in memory
+    Low-level class responsible for storing and retrieving calculated vectors in memory.
     """
 
-    def __init__(self, is_eternal = False):
+    _arrays: dict
+    is_eternal: bool
+
+    def __init__(self, is_eternal: bool = False) -> None:
         self._arrays = {}
         self.is_eternal = is_eternal
 
-    def get(self, period):
+    def get(self, period: periods.Period) -> numpy.ndarray:
         if self.is_eternal:
-            period = periods.period(ETERNITY)
-        period = periods.period(period)
+            period = periods.period(periods.ETERNITY)
 
+        period = periods.period(period)
         values = self._arrays.get(period)
+
         if values is None:
             return None
+
         return values
 
     def put(self, value, period):
         if self.is_eternal:
-            period = periods.period(ETERNITY)
-        period = periods.period(period)
+            period = periods.period(periods.ETERNITY)
 
+        period = periods.period(period)
         self._arrays[period] = value
 
-    def delete(self, period = None):
+    def delete(self, period: Optional[periods.Period] = None) -> None:
         if period is None:
             self._arrays = {}
             return
 
         if self.is_eternal:
-            period = periods.period(ETERNITY)
+            period = periods.period(periods.ETERNITY)
+
         period = periods.period(period)
 
-        self._arrays = {
-            period_item: value
-            for period_item, value in self._arrays.items()
-            if not period.contains(period_item)
-            }
+        if period is not None:
+            self._arrays = self._pop(period, list(self._arrays.items()))
 
-    def get_known_periods(self):
-        return self._arrays.keys()
+    def get_known_periods(self) -> List[periods.Period]:
+        return list(self._arrays.keys())
```

Comment:
Be careful here: this is a change of behaviour. `dict.keys()` returns a (dynamic) view, the `list()` calls freezes it. (The change is not necessary, as this was already being frozen by the calling code in Holder.)

I don't think the change is consequential, but it makes the title of the PR a lie, as we're adding tests *and* changing behaviour - *and* changing the CI config *and* adding type annotations *and* changing the style of imports… PRs that do many things at once are harder to review and more likely to contain unnoticed errors.

Reply:
> We're adding tests

Check :)

> Changing behaviour

Should be avoided, thanks for the pointer.

> Changing the CI config

Was randomly changing

> Adding type annotations

That's documentation.

As I understand this project already adopts type annotations, and a progressive strategy to annotate code. I'm just doing what the linter does: check as I go.

I feel you don't like them, and that's fine. Just take one thing into consideration:

- I don't feel the current storage implementation is  good one
- In order to change it, I'm trying to get a sense of their behaviours
- so I can later move things around with easy, for example to provide an explicit cache API
- I type annotate before writing tests as contract hypothesis that I correct after testing

> Changing the style of imports

Should be avoided, thanks for the pointer.

> PRs that do many things at once are harder to review and more likely to contain unnoticed errors.

I'll try to reduce the noise, thanks !

Reply:
> I feel you don't like them

Nah. I love explicit types, otherwise I wouldn't be doing so much Haskell at the moment :)

Reply:
❤️ tried to remove as much noise as I can! :)